csharp/src/Apache.Arrow.Adbc/Extensions/IArrowArrayExtensions.cs

```diff
@@ -198,7 +208,23 @@ public static class IArrowArrayExtensions
                 case ArrowTypeId.Int64:
                     return (array, index) => ((Int64Array)array).GetValue(index);
                 case ArrowTypeId.String:
-                    return (array, index) => ((StringArray)array).GetString(index);
+                    return (array, index) =>
+                    {
+                        StringArray? sArray = array as StringArray;
```

Comment:
This gives up some of the performance benefit of the approach. Instead of having to add a check to each invocation, can we have the caller tell us in advance what the expected source type is and return one of two different delegates?

Reply:
I couldnt figure out an elegant way to have two separate delegates so I went with one and checking the DataType of the array that's passed in.

Reply:
It's still a shame to have a branch on the critical path -- especially for a such a commonly-used data type. At least in this case, the CPU's branch predictor ought to figure out pretty quickly that the array is almost never going to be a decimal256. And I can see that making things work that way would require a fair amount of refactoring.